markdown: Add `hl_lines` support to Markdown Processor.

akshatdalton opened the pull request:
This commit adds the new `hl_lines` feature of CodeHilite which helps to tag certain lines of a codeblock to be highlighted for demonstration purposes.

Line numbers, say 'n1' and 'n2', can be tagged as hl_lines="n1 n2".
Index in codeblock starts from 1.

Fixes #2036.

**GIFs or screenshots:** 
## Source:
~~~
```ruby hl_lines="1 3 5"
# Simple for loop using a range.
for i in (1..4)
  print i," "
end
print "\n"
```
~~~
## Result:
![Screenshot from 2020-12-30 15-39-24](https://user-images.githubusercontent.com/67277428/103344754-84a6ec00-4ab5-11eb-935b-ae5edf137994.png)

akshatdalton commented:
Should we also add a type-ahead for `hl_lines=""` ?

andersk commented:
Upstream CodeHilite does not support ranges, and we shouldn’t either, both for compatibility, and because someone might DoS the server by making it expand a range into a list of trillions of integers.

timabbott commented on zerver/lib/markdown/fenced_code.py:
```diff
@@ -237,19 +248,35 @@ def done(self) -> None:
+
+    try:
+        return list(map(int, hl_lines.split()))
+    except ValueError:  # nocoverage
```
Can we add a test covering this?

timabbott replied:
Or if we think that's impossible, maybe we should make it a `raise AssertionError` instead?

timabbott commented:
Thanks for working on this, and sorry for the slow review!  This is nice work.  I posted one stream of comments; otherwise this lgtm.  

If you're looking for a next project, one of the oldest open issues is re-enabling backtick escaping in Zulip's markdown processors, and that'd be well worth doing.

timabbott commented:
OK; I think the main thing we still need to do is documentation; I'll think about that and then hopefully merge.

timabbott commented:
Oh, actually, can you post a night theme screenshot?  My guess is we need some CSS for this to look OK in that theme.

You might as well also document this feature in `templates/zerver/help/format-your-message-using-markdown.md` since we're doing another review round trip.

akshatdalton commented:
@timabbott, yes, you are right. This need some CSS work for the night theme.
| Day Theme | Night Theme |
| --- | --- |
| ![hl_lines_day](https://user-images.githubusercontent.com/67277428/109148080-c9849180-778b-11eb-93bb-86c94cc8892a.png) | ![hl_lines_night](https://user-images.githubusercontent.com/67277428/109148127-d5705380-778b-11eb-952d-5e8b89a39ccd.png) |

In both of them, same line numbers are tagged (1, 3 and 6).

Following is the screenshot of the updated user documentation.

### User Documentation
![Codehilite_documentation](https://user-images.githubusercontent.com/67277428/109148922-e077b380-778c-11eb-9829-e6a6e1331701.png)

### Message Formatting
![message_formatting](https://user-images.githubusercontent.com/67277428/109153459-f12b2800-7792-11eb-994d-b0b9f13dd46d.png)

timabbott commented:
I find it really hard to see in the night theme screenshots; maybe we should google around for "pygments dark css" and see if there's a standard color scheme we should use.  Also, I am used to the whole line being highlighted, not just the part that has code; that should look better.

https://gohugo.io/content-management/syntax-highlighting/#example-highlight-shortcode

akshatdalton commented:
@timabbott After making the whole line highlighted the following dark version looks better to me.
### Dark Theme
![hl_lines_dark_new](https://user-images.githubusercontent.com/67277428/109212078-193b7b00-77d5-11eb-8ab3-7ff81aadb9dc.png)
### Day Theme
![hl_lines_day_new](https://user-images.githubusercontent.com/67277428/109212387-79cab800-77d5-11eb-8e5b-5ebcb8241e50.png)
---------------------
![Screenshot from 2021-02-26 01-57-25](https://user-images.githubusercontent.com/67277428/109212974-17be8280-77d6-11eb-9b39-ca57d8da84c4.png)
Any thoughts ? Still want to look around for some other "pygments dark css" ?

akshatdalton commented on zerver/lib/markdown/fenced_code.py:
```diff
@@ -237,19 +250,37 @@ def done(self) -> None:
+    if not hl_lines_str:
+        return []
+
+    try:
+        return list(map(int, hl_lines_str.split()))
+    except AssertionError:  # nocoverage
```
@timabbott @andersk Should I write it as -
```python
except ValueError as e: # nocoverage
  raise AssertionError(e)
```
or as -
```python
except ValueError:
  return []
```
Latter is the upstream version.

timabbott replied:
I think Anders' proposal was that we call the upstream function directly rather than copying it, but if that's not possible, I'd prefer to copy it exactly with a comment.

akshatdalton commented on zerver/lib/markdown/fenced_code.py:
```diff
@@ -109,6 +109,8 @@
     ) # language, like ".py" or "{javascript}"
     [ ]* # spaces
     (
+        (hl_lines=(?P<quot>"|')(?P<hl_lines>[0-9 ]*)(?P=quot))?
+        | # hl_lines to emphasize certain lines of code
         \\{?\\.?
         (?P<header>
             [^~`]*
```
@andersk 
Ok, I will update it as -
```python
    (
        (
            \\{?\\.?
            (hl_lines=(?P<quot>"|')
                (?P<hl_lines>[0-9 ]*)(?P=quot)
            )
            \\}?
        ) # hl_lines to emphasize certain lines of code
        |
        (
            \\{?\\.?
            (?P<header>
                [^~`]*
            )
            \\}?
        ) # header for features that use fenced block header syntax (like spoilers)
    )
```
To make it more explicate ?

andersk replied:
That’s even more convoluted. What are the extra `(`, `\\{?\\.?`, `(`, `)`, `\\}?`, `)`, `(`, `)` for?

akshatdalton replied:
Okay, may be I understood this wrong. Can you please give an example how: `the empty string ambiguously matches either alternative` ? Does that mean for the initial modified regex, it creates problem for the following code ? -
```
```python
def zulip():
  print("Zulip")
```

The regex:
```
(hl_lines=(?P<quot>"|')(?P<hl_lines>[0-9 ]*)(?P=quot))?
```
is picked from the upstream [`markdown.extensions.codehilite._parseHeader`](https://github.com/Python-Markdown/markdown/blob/14c2fa92302084e4850e5323ed56721e43fbbdb9/markdown/extensions/codehilite.py#L195)

timabbott commented on templates/zerver/help/format-your-message-using-markdown.md:
`````diff
@@ -118,6 +119,25 @@ Organization administrators can also configure a default syntax
 highlighting language.  In this configuration, one can use ````text`
 to display content without any syntax highlighting.
 
+## Codehilite
`````
Also, `codehilite` refers to syntax highlighting, not line numbers.  So "line highlighting" as a subsection seems appropriate here.

timabbott commented:
I still find the highlighting colors you have in the dark theme basically invisible.  Maybe post on #design on chat.zulip.org for help picking a color.  If I were to do it, I'd probably try something in the slightly yellow direction.  You can also try checking products like GitHub or GitLab that might have a dark theme and code highlighting features.

akshatdalton commented:
I have updated the color use for line highlighting in dark mode to that one use by GitLab. Moreover, this now highlights the complete line. Following are the screenshots:
## Day Mode
![Day Mode](https://user-images.githubusercontent.com/67277428/111181380-eca8a100-85d3-11eb-94e8-b26ae81fae38.png)
## Night Mode
![Night Mode](https://user-images.githubusercontent.com/67277428/111181418-f7fbcc80-85d3-11eb-9dc4-7d578ec22e0e.png)

Though there is a small problem, in `Message formatting` section, the example code is rendering as:
![Message formatting](https://user-images.githubusercontent.com/67277428/111181912-70fb2400-85d4-11eb-9793-2a2d7dd24fe6.png)

An extra empty line is added, that's probably because I have added `display: block` property. I tried other properties but couldn't get the desired result.

@timabbott please review.

timabbott commented:
I think this is basically ready to merge from a UX perspective.  The regular expression still needs work (https://github.com/zulip/zulip/pull/16984#discussion_r583172886) though.

timabbott commented:
I opened issues with the terminal and mobile apps to make sure those projects get tested against this.

akshatdalton committed:
markdown: Add `hl_lines` support to Markdown Processor.

This commit adds the support of `hl_lines` feature of CodeHilite
which helps to tag certain lines of a codeblock to be highlighted
for demonstration purposes.

Line numbers, say 'n1', 'n2', of a codeblock can be tagged as

```<lang> hl_lines="n1 n2".
// code block content
```

Where <lang> can be ruby, .py, {javascript}, etc.
Index in codeblock starts from 1.

Range of line numbers, n1-n2, is not supported to prevent DoS
attacks as one can expand the range into a list of trillions of
numbers.

Fixes #2036.

akshatdalton committed:
docs: Add documentation for `hl_lines` feature of Codehilite.

The documentation for the `hl_lines` feature of Codehilite, aka
`Line Highlighting`, is added as a subsection of `Code blocks`.

The Message formatting section (under Keyboard shortcuts)
is updated as well.

timabbott commented:
@akshatdalton what's the status of fixing the regular expression in this PR?

akshatdalton commented:
@timabbott, I started a thread: https://chat.zulip.org/#narrow/stream/3-backend/topic/.60hl_lines.60.20regex.20.232036, and I have given my explanation there. I want @andersk to confirm that new changes don't lead to any empty string match ambiguously [either alternative](https://github.com/zulip/zulip/pull/16984#discussion_r583001775).

timabbott commented on static/js/info_overlay.js:
```diff
@@ -98,6 +98,16 @@ def zulip():
+    print "Zulip"
+\`\`\``,
+        output_html: `\
+<div class="codehilite"><pre><span class="hll"><span class="k">def</span> <span class="nf">zulip</span><span class="p">():</span></span>
```
Remove this; we don't need this advanced feature in the cheat sheet.

timabbott commented on templates/zerver/help/format-your-message-using-markdown.md:
````diff
@@ -118,6 +118,22 @@ Organization administrators can also configure a default syntax
+```
+~~~
+
+![Markdown line highlighting](/static/images/help/markdown-line-highlighting.png)
````
Let's move this to the new /help/code-blocks.
